server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,308 @@
+                        PublicationTransportHandler.PUBLISH_STATE_ACTION_NAME,
+                        // This disables pre-voting on all nodes except the new master, forcing it to win the election
+                        StatefulPreVoteCollector.REQUEST_PRE_VOTE_ACTION_NAME
+                    );
+                    blockActionNameOnMockTransportService(mockTransportService, listOfActionsToBlock);
+                }
+            }
+
+            // We do not expect to see a WARN log about a node disconnecting (#ES-11449)
+            addJoiningNodeDisconnectedWarnLogFalseExpectation(mockLog);
+
+            // We haven't changed master nodes yet
+            assertEquals(originalMasterNodeName, internalCluster().getMasterName());
+
+            // Sends a node join request to the original master node. This will fail, and cause a master failover
+            // startDataOnlyNode waits for the new node to be added, and this can only occur after a re-election
+            String newNodeName = internalCluster().startDataOnlyNode();
+            assertNotEquals(originalMasterNodeName, internalCluster().getMasterName());
+            logger.info("New master is elected");
+
+            mockLog.assertAllExpectationsMatched();
+
+            // Assert the new data node was added
+            DiscoveryNodes discoveryNodes = internalCluster().clusterService().state().nodes();
+            assertEquals(numberOfNodesOriginallyInCluster + 1, discoveryNodes.getSize());
+            assertEquals(numberOfDataNodesOriginallyInCluster + 1, discoveryNodes.getDataNodes().size());
+            assertTrue(getListOfDataNodeNamesFromCluster(newMasterNodeName).contains(newNodeName));
+            assertEquals(numberOfMasterNodesOriginallyInCluster, discoveryNodes.getMasterNodes().size());
+        }
+    }
+
+    /*
+        In this scenario, node N attempts to join a cluster, there is an election and the original master is re-elected.
+        Node N should join the cluster, but it should not be disconnected (#ES-11449)
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.cluster.coordination:INFO")
+    public void testNodeTriesToJoinClusterAndThenSameMasterIsElected() {
+        internalCluster().startNodes(3);
+        ensureStableCluster(3);
+        String masterNodeName = internalCluster().getMasterName();
+
+        long originalTerm = internalCluster().clusterService(masterNodeName).state().coordinationMetadata().term();
+        int numberOfNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName).state().getNodes().size();
+        int numberOfMasterNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName)
+            .state()
+            .nodes()
+            .getMasterNodes()
+            .size();
+        int numberOfDataNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName).state().nodes().getDataNodes().size();
```

Comment:
Likewise here, I think we just need to count the nodes, no need to distinguish them by role.